### PR TITLE
bot: replaced retry function with Tenacity

### LINKED ISSUE #110: Use tenacity (or similar) instead of re-implementing a retry function
We are currently implementing https://github.com/mozilla/code-coverage/blob/10285fb2340ca69343503b9b42cffebd878f2bae/bot/code_coverage_bot/utils.py#L20 and using it in a few places.
We should instead use a library (e.g. tenacity) and remove this custom `retry` function.

### PR DESCRIPTION
**Pull request by seb-sojka**
Remove retry function and related RunException from utils.py for bot and replaces it with Tenacity retry in taskcluster.py and uploader.py
Fixes #110

### CODE REVIEW

File: bot/code_coverage_bot/taskcluster.py [hunk 1]
```diff
@@ -98,8 +99,6 @@ def perform_download():
         if artifact_path.endswith(".zip") and not is_zipfile(artifact_path):
             raise BadZipFile("File is not a zip file")
 
-    retry(perform_download)
```
Comment (marco-c) [inline bot/code_coverage_bot/taskcluster.py]: You should call the function here :P

File: bot/code_coverage_bot/taskcluster.py [hunk 2]
```diff
@@ -87,6 +87,7 @@ def download_artifact(artifact_path, task_id, artifact_name):
     url = queue.buildUrl("getLatestArtifact", task_id, artifact_name)
     logger.debug("Downloading artifact", url=url)
 
+    @retry(wait=wait_fixed(30), stop=stop_after_attempt(5))
```
Comment (marco-c) [inline bot/code_coverage_bot/taskcluster.py]: What would be the defaults?

**Comment by marco-c** [comment]
There are some linting failures (I suggest using pre-commit for that), and some test failures (you can run tests by doing `cd bot` and then `python -m pytest -s -v tests/`).

File: bot/requirements-dev.txt [hunk 3]
```diff
@@ -4,3 +4,4 @@ pre-commit==1.21.0
 pytest==5.3.2
 pytest-responses==0.4.0
 responses==0.10.9
+tenacity==6.0.0
```
Comment (marco-c) [inline bot/requirements-dev.txt]: It should not be in requirements-dev.txt, but in requirements.txt, since it is used in production code and not only in tests.

File: bot/code_coverage_bot/taskcluster.py (hunk 2 above)
Comment (seb-sojka) [reply bot/code_coverage_bot/taskcluster.py]: It would retry forever without waiting without the arguments. The numbers were chossen was they were the defaults used in retry function.

Comment (marco-c) [reply bot/code_coverage_bot/taskcluster.py]: OK, sounds good!

**Comment by marco-c** [comment]
@seb-sojka do you still want to work on this?

File: bot/requirements-dev.txt [hunk 4]
```diff
@@ -3,4 +3,4 @@ jsonschema==3.2.0
 pre-commit==1.21.0
 pytest==5.3.2
 pytest-responses==0.4.0
-responses==0.10.9
+responses==0.10.9
```
Comment (marco-c) [inline bot/requirements-dev.txt]: Please revert the changes in this file, since they are unrelated.

File: bot/code_coverage_bot/uploader.py [hunk 5]
```diff
@@ -45,10 +46,10 @@ def gcp(repository, revision, report, platform, suite):
...
 
     # Trigger ingestion on backend
-    retry(
-        lambda: gcp_ingest(repository, revision, platform, suite),
-        retries=10,
-        wait_between_retries=60,
+    @tenacity.retry(
+        retry = retry_if_exception(lambda: gcp_ingest(repository, revision, platform, suite)),
+        stop=stop_after_attempt(10),
+        wait=wait_fixed(60)
```
Comment (marco-c) [inline bot/code_coverage_bot/uploader.py]: Nit: linting here will probably complain, as you are using ` = ` once and `=` without spaces twice. Use pre-commit to format everything.

**Comment by seb-sojka** [comment]
Hello,
I think have run into an issue with the tester with changing to Tenacity for retries for perform_download. It does not increase the call count but I have used print to see the perform_download occur 5 times.

File: bot/code_coverage_bot/utils.py [hunk 6]
```diff
@@ -17,21 +16,6 @@ class RunException(Exception):
     """
```
Comment (marco-c) [inline bot/code_coverage_bot/utils.py]: You can remove `RunException` at this point, and just use `Exception` in `run_check`

**Comment by marco-c** [comment]
>  I have run into an issue with the tester when changing to Tenacity for retries for perform_download function. It does not increase the call count but I have used print to see the perform_download occur 5 times.

Probably tenacity is not using time.sleep. You can change the test to stop mocking it, and instead count the number of times the responses URL is hit (see https://github.com/getsentry/responses, something like `assert len(responses.calls) == 5`).

**Comment by seb-sojka** [comment]
I failing linit test because of length of line I did not change and outside the scope of the request. Should I make changes to past the test or something else?

**Comment by seb-sojka** [comment]
The error with linit and taskcluster_yml. I am unsure what the solution for this. I do not have this error locally when I run the pre-commit.

File: bot/code_coverage_bot/uploader.py [hunk 7]
```diff
@@ -66,6 +62,7 @@ def gcp_covdir_exists(repository, revision, platform, suite):
     return blob.exists()
 
 
+@tenacity.retry(stop=tenacity.stop_after_attempt(10), wait=tenacity.wait_fixed(60))
```
Comment (marco-c) [inline bot/code_coverage_bot/uploader.py]: Use reraise=True here as well.

**Comment by marco-c** [comment]
Looks good, just one last thing to fix!

**Comment by marco-c** [comment]
If you want, here are some similar issues in this and other repos:
#392 
https://github.com/mozilla/bugbug/issues/1234
